File: file-templates/templates/MVVM/List.xctemplate/UIViewObjective-C/___FILEBASENAME___View.m
//___FILEHEADER___

#import "___FILEBASENAME___.h"
#import "___VARIABLE_productName___ViewModel.h"
#import <Masonry/Masonry.h>

@interface ___VARIABLE_productName:identifier___View ()<UITableViewDataSource, UITableViewDelegate>
@property (nonatomic, strong) ___VARIABLE_productName___ViewModel *viewModel;
@property (nonatomic, strong) UIView *contentView;
@property (nonatomic, strong) UITableView *tableView;
@end

@implementation ___VARIABLE_productName___View

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        
    }
    return self;
}

- (instancetype)initWithViewModel:(___VARIABLE_productName___ViewModel *)viewModel {
    self = [super init];
    if (self) {
        self.viewModel = viewModel;
    }
    return self;
}

#pragma mark - private

- (void)onUpdateUIComponents {
    // TODO: implement
}

- (void)onSetuBinding {
    [self onUpdateUIComponents];
    // binding implement
}

#pragma mark - Public

- (void)reloadData {
    [self.tableView reloadData];
}

#pragma mark - UITableViewDataSource & UITableViewDelegate

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return [self.viewModel numberOfSections];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.viewModel numberOfRowsInSection:section];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    id item = [self.viewModel itemForRowAtIndexPath:indexPath];
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"" forIndexPath:indexPath];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    [self.viewModel onSelectRowAtIndexPath:indexPath];
}

#pragma mark - getter/setter

- (void)setViewModel:(___VARIABLE_productName___ViewModel *)viewModel {
    if (_viewModel != viewModel) {
        _viewModel = viewModel;
        [self onSetuBinding];
    }
}

- (UIView *)contentView {
    if (!_contentView) {
        UIView *view = [[UIView alloc] initWithFrame:CGRectZero];
        [self addSubview:view];
        [view mas_makeConstraints:^(MASConstraintMaker *make) {
            make.edges.inset(0);
        }];
        _contentView = view;
    }
    return _contentView;
}


- (UITableView *)tableView {
    if (!_tableView) {
        UITableView *view = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStylePlain];
        view.dataSource = self;
        view.delegate = self;
        view.backgroundColor = [UIColor whiteColor];
        view.showsVerticalScrollIndicator   = NO;
        view.showsHorizontalScrollIndicator = NO;
        view.separatorStyle = UITableViewCellSeparatorStyleNone;
        view.rowHeight = 44;
        [view registerClass:[UITableViewCell class] forCellReuseIdentifier:@""];
        
        if(@available(iOS 11.0, *)) {
            view.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentNever;
        }
        if (@available(iOS 15.0, *)) {
            view.sectionHeaderTopPadding = 0;
        } else {
            // Fallback on earlier versions
        }
        [self.contentView addSubview:view];
        [view mas_makeConstraints:^(MASConstraintMaker *make) {
            make.edges.inset(0);
        }];
        _tableView = view;
    }
    return _tableView;
}

@end

File: file-templates/templates/MVVM/TableList.xctemplate/UIViewControllerSwift/___FILEBASENAME___View.swift
//  ___FILEHEADER___
//  Summary: ___VARIABLE_Summary___

import UIKit
import SnapKit

final class ___VARIABLE_productName:identifier___View: UIView {
    
    // MARK: - Private properties
    
    private var viewModel: ___VARIABLE_productName:identifier___ViewModel {
        didSet {
            onDataBindings()
        }
    }

    private lazy var contentView: UIView = {
        let view = UIView()
        view.backgroundColor = .clear
        return view
    }()
    
    private lazy var tableView: UITableView = {
        let tableView = UITableView(frame: .zero, style: .plain)
        tableView.dataSource = self
        tableView.delegate = self
        tableView.backgroundColor = .clear
        tableView.separatorStyle = .singleLine
        tableView.rowHeight = 48
        if #available(iOS 15.0, *) {
            tableView.sectionHeaderTopPadding = 0
        } else {
            // Fallback on earlier versions
        }
        tableView.separatorInset = UIEdgeInsets(top:0, left:0, bottom: 0, right: 0)
        tableView.contentInset = UIEdgeInsets(top:0, left:0, bottom: 0, right:0)
        tableView.separatorColor = UIColor(red: 219/255.0, green: 219/255.0, blue: 219/255.0, alpha: 1)
        tableView.register(___VARIABLE_productName___Cell.self, forCellReuseIdentifier:"___VARIABLE_productName___Cell")
//        tableView.registerClass(___VARIABLE_productName___Cell.self)
        return tableView
    }()

    // MARK: - Init

    init(viewModel: ___VARIABLE_productName:identifier___ViewModel) {
        self.viewModel = viewModel
        super.init(frame: .zero)
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: - UIComponents

    func onInitUIComponents() {
        addSubview(contentView)
        contentView.addSubview(tableView)
    }
    
    func onLayoutUIComponents() {
        contentView.snp.makeConstraints { make in
            make.edges.equalTo(UIEdgeInsets(top: 0, left: 0, bottom: 0, right: 0))
        }

        tableView.snp.makeConstraints { make in
            make.edges.equalToSuperview()
        }
    }
    
    func onUpdateUIComponents() {

    }

    func onDataBindings() {
        onUpdateUIComponents()
        self.viewModel.reloadSubject.subscribe { [weak self] ret in
            guard let self = self else { return }
            self.reloadData()
        }.disposed(by: disposeBag)
    }
}

// MARK: - Public

extension ___VARIABLE_productName:identifier___View {

    func reloadData() {
        onUpdateUIComponents()
        self.tableView.reloadData()
//        DispatchQueue.main.dispatchMainIfNeeded { [weak self] in
//            guard let self = self else { return }
//            
//        }
    }
}

// MARK: - UITableViewDataSource & UITableViewDelegate

extension ___VARIABLE_productName:identifier___View: UITableViewDataSource, UITableViewDelegate {

    func numberOfSections(in tableView: UITableView) -> Int {
        return viewModel.numberOfSections()
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return viewModel.numberOfRowsInSection(section)
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "___VARIABLE_productName___Cell", for: indexPath) as! ___VARIABLE_productName___Cell
        cell.selectionStyle = .none
        guard let cellModel = viewModel.itemForRowAt(indexPath) else {
            return cell
        }
        
        cell.cellModel = cellModel
        return cell
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return viewModel.heightForRowAt(indexPath)
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        viewModel.onDidSelectRowAt(indexPath)
    }

}
